### PR TITLE
coverage: Use upstream coverage collect script

### PR DESCRIPTION
**Pull request by phlax**
Our custom collect script should no longer be necessary as upstream now has the support for which it was added

This speeds coverage up by ~25% in my testing, and uses a lot less disk space

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: bazel/coverage/collect_cc_coverage.sh [hunk 1]
```diff
@@ -63,31 +63,49 @@ function init_gcov() {
...
-  shift
+function llvm_coverage_lcov() {
+  local output_file="${1}"; shift
   export LLVM_PROFILE_FILE="${COVERAGE_DIR}/%h-%p-%m.profraw"
+
   "${COVERAGE_GCOV_PATH}" merge -output "${output_file}.data" \
       "${COVERAGE_DIR}"/*.profraw
 
+  # TODO(phlax): figure out why this is necessary, and try to switch to upstream
+  RUNFILES_DIR_PREFIX="${RUNFILES_DIR}/envoy/"
```
Comment (phlax) [inline bazel/coverage/collect_cc_coverage.sh:88]: added hack

**Comment by phlax** [comment]
it might be better to use a patch rather than custom file - best would be to not need any hacks

File: bazel/coverage/collect_cc_coverage.sh [hunk 2]
```diff
@@ -63,31 +63,49 @@ function init_gcov() {
...
 
-  object_files="$(find -L "${RUNFILES_DIR}" -type f -exec file -L {} \; \
-       | grep ELF | grep -v "LSB core" | sed 's,:.*,,')"
+  local object_param=""
```
Comment (phlax) [inline bazel/coverage/collect_cc_coverage.sh:90]: current implementation here was matching a ton of files it should not

File: bazel/coverage/collect_cc_coverage.sh [hunk 3]
```diff
@@ -134,17 +164,36 @@ function gcov_coverage() {
...
+              # Delete the .gcov files.
+              rm -- *.gcov
+          fi
       fi
     fi
   done < "${COVERAGE_MANIFEST}"
 }
 
+BAZEL_CC_COVERAGE_TOOL="LLVM_LCOV"
+GENERATE_LLVM_LCOV=1
```
Comment (phlax) [inline bazel/coverage/collect_cc_coverage.sh:195]: rm

File: .azure-pipelines/pipelines.yml [hunk 4]
```diff
@@ -62,6 +62,9 @@ stages:
 # Presubmit/default
 - ${{ if eq(variables.pipelineDefault, true) }}:
   - template: stages.yml
+    parameters:
+      checkStageDeps:
+      - env
```
Comment (phlax) [inline .azure-pipelines/pipelines.yml:67]: rm

File: test/run_envoy_bazel_coverage.sh [hunk 5]
```diff
@@ -74,14 +74,22 @@ BAZEL_OUTPUT_BASE="$(bazel "${BAZEL_STARTUP_OPTIONS[@]}" info "${BAZEL_BUILD_OPT
 echo "Running bazel coverage with:"
 echo "  Options: ${BAZEL_BUILD_OPTIONS[*]} ${BAZEL_COVERAGE_OPTIONS[*]}"
 echo "  Targets: ${COVERAGE_TARGETS[*]}"
-bazel "${BAZEL_STARTUP_OPTIONS[@]}" coverage "${BAZEL_BUILD_OPTIONS[@]}" "${BAZEL_COVERAGE_OPTIONS[@]}" "${COVERAGE_TARGETS[@]}"
+
+# CANONICAL_FLAGS="$(bazel "${BAZEL_STARTUP_OPTIONS[@]}" canonicalize-flags "${BAZEL_BUILD_OPTIONS[@]}" "${BAZEL_COVERAGE_OPTIONS[@]}")"
+
+# echo "CANONICAL: X${CANONICAL_FLAGS}X"
```
Comment (phlax) [inline test/run_envoy_bazel_coverage.sh:80]: rm

File: test/run_envoy_bazel_coverage.sh [hunk 6]
```diff
@@ -108,6 +116,15 @@ else
     cp bazel-out/_coverage/_coverage_report.dat "${COVERAGE_DATA}"
 fi
 
+# This worksaround issue in fuzzcov - possibly better solution involves instrumentation_filter
```
Comment (phlax) [inline test/run_envoy_bazel_coverage.sh:119]: this seems to be necessary

File: test/run_envoy_bazel_coverage.sh [hunk 7]
```diff
@@ -72,14 +72,22 @@ BAZEL_OUTPUT_BASE="$(bazel "${BAZEL_STARTUP_OPTIONS[@]}" info "${BAZEL_BUILD_OPT
 echo "Running bazel coverage with:"
 echo "  Options: ${BAZEL_BUILD_OPTIONS[*]} ${BAZEL_COVERAGE_OPTIONS[*]}"
 echo "  Targets: ${COVERAGE_TARGETS[*]}"
-bazel "${BAZEL_STARTUP_OPTIONS[@]}" coverage "${BAZEL_BUILD_OPTIONS[@]}" "${BAZEL_COVERAGE_OPTIONS[@]}" "${COVERAGE_TARGETS[@]}"
+
+# CANONICAL_FLAGS="$(bazel "${BAZEL_STARTUP_OPTIONS[@]}" canonicalize-flags "${BAZEL_BUILD_OPTIONS[@]}" "${BAZEL_COVERAGE_OPTIONS[@]}")"
+
+# echo "CANONICAL: X${CANONICAL_FLAGS}X"
+
+bazel "${BAZEL_STARTUP_OPTIONS[@]}" coverage --announce_rc "${BAZEL_BUILD_OPTIONS[@]}" "${BAZEL_COVERAGE_OPTIONS[@]}" "${COVERAGE_TARGETS[@]}"
```
Comment (phlax) [inline test/run_envoy_bazel_coverage.sh:80]:
```suggestion
bazel "${BAZEL_STARTUP_OPTIONS[@]}" coverage "${BAZEL_BUILD_OPTIONS[@]}" "${BAZEL_COVERAGE_OPTIONS[@]}" "${COVERAGE_TARGETS[@]}"
````

**Comment by phlax** [comment]
this speeds coverage up pretty dramatically

image

more importantly its a step towards enabling rbe for coverage which should hopefully speed up more

File: test/run_envoy_bazel_coverage.sh [hunk 8]
```diff
@@ -106,6 +110,15 @@ else
     cp bazel-out/_coverage/_coverage_report.dat "${COVERAGE_DATA}"
 fi
 
+# This worksaround https://github.com/bazelbuild/bazel/issues/19237
+echo "remove external files from coverage report"
+lcov -r "${COVERAGE_DATA}" "*external*" -o "${COVERAGE_DATA}.new"
```
Comment (phlax) [inline test/run_envoy_bazel_coverage.sh:115]:
```suggestion
lcov -r "${COVERAGE_DATA}" "/external*" -o "${COVERAGE_DATA}.new"
```

this is removing mobile files it should not

Comment (phlax) [reply test/run_envoy_bazel_coverage.sh:115]: hack not required https://github.com/bazelbuild/bazel/issues/19237#issuecomment-1675999124

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**